File: backend/data/ETL/pipeline.py
"""
    The whole ETL in one call: Excel file in, products in the database out.
    Used by the frontend uploader.
"""
from backend.data.relationaldb import apply_schema, get_db_connection
from backend.data.ETL.Extraction.excelDataExtraction import read_rows
from backend.data.ETL.Transformation.normalizeProducts import transform
from backend.data.ETL.Load.loadProducts import load


def import_excel(path, conn=None):
    """
        Returns (inserted, updated, skipped). Raises ValueError if the workbook
        has no rows this system can recognise as products.
    """
    products, skipped = transform(read_rows(path))
    if not products:
        raise ValueError("no_products")  # key into backend.i18n.STRINGS, translated by the caller

    own_connection = conn is None
    conn = conn or get_db_connection()
    try:
        inserted, updated = load(apply_schema(conn), products)
    finally:
        if own_connection:
            conn.close()
    return inserted, updated, skipped

File: src/main/java/org/example/etl/Pipeline.java
package org.example.etl;

import org.example.db.Database;
import org.example.model.Product;

import java.io.IOException;
import java.nio.file.Path;
import java.sql.Connection;
import java.sql.SQLException;
import java.util.List;

/**
 * The whole ETL in one call: Excel file in, products in the database out.
 * Used by the UI uploader.
 */
public final class Pipeline {
    private Pipeline() {
    }

    public record Result(int inserted, int updated, int skipped) {
    }

    /** Uses its own connection, opened and closed here. */
    public static Result importExcel(Path path) throws IOException, SQLException {
        try (Connection conn = Database.connect()) {
            return importExcel(path, conn);
        }
    }

    /**
     * Reuses the given connection (the caller owns closing it). Throws
     * IllegalArgumentException("no_products") -- a key into Messages, translated
     * by the caller -- if the workbook has no rows this system can recognise as
     * products.
     */
    public static Result importExcel(Path path, Connection conn) throws IOException, SQLException {
        List<java.util.Map<String, String>> rawRows = ExcelReader.readRows(path);
        ProductNormalizer.Result transformed = ProductNormalizer.transform(rawRows);
        List<Product> products = transformed.products();
        if (products.isEmpty()) {
            throw new IllegalArgumentException("no_products");
        }

        Database.applySchema(conn);
        ProductLoader.Result loaded = ProductLoader.load(conn, products);
        return new Result(loaded.inserted(), loaded.updated(), transformed.skipped());
    }
}
